assets/app/features/upgrade/steps.factory.spec.js

```diff
@@ -118,5 +406,87 @@ describe('Stepes Factory', function () {
+                expect(upgradeStepsFactory.stepByID(9999)).not.toBeDefined();
+            });
+        });
+
+        describe('lastStateForRestore function', function () {
+            describe('when there is no stored last state', function () {
+                beforeEach(function () {
+                    localStorage.removeItem(UPGRADE_LAST_STATE_KEY);
+                });
+                it('should return "upgrade-landing" before any step', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(initialStatusData))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" after prechecks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrechecks))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade-landing" during prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringPrepare))
+                        .toEqual('upgrade-landing');
+                });
+                it('should return "upgrade.backup" after prepare', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterPrepare))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" during backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataDuringBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.backup" after backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterBackup))
+                        .toEqual('upgrade.backup');
+                });
+                it('should return "upgrade.administration-repository-checks" after admin repo checks', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterAdminRepoChecks))
+                        .toEqual('upgrade.administration-repository-checks');
+                });
+            });
+            describe('when stored state is invalid in given case', function () {
+                beforeEach(function () {
+                    localStorage.setItem(UPGRADE_LAST_STATE_KEY, 'upgrade-landing');
+                });
+                it('should return "upgrade.backup" after backup', function() {
+                    expect(upgradeStepsFactory.lastStateForRestore(statusDataAfterBackup))
+                        .toEqual('upgrade.backup');
+                });
+            });
+            describe('when stored state points to current, pending step', function () {
+                beforeEach(function () {
+                    localStorage.setItem(UPGRADE_LAST_STATE_KEY, 'upgrade.administration-repository-checks');
```

Comment:
`upgrade-landing` is a valid state, not step (which is confusing enough, but still valid for that case), but it doesn't validate `lastStateForRestore` method returns when a different (yet valid) state is stored in localstorage.
A case for this could be if the user decides to open a different browser to continue the upgrade flow some steps and then refresh the original one, or any error that could've been introduced in `showNextStep` or `cancelUpgrade` methods in the factory and upgrade.controller